### PR TITLE
Revise reference implementation for ReduceMean operation

### PR DESCRIPTION
**Pull request by ggalieroc**
Details: ReduceMean reference implementation refactored.

- Remove CoordinateTransform calls to index(). Benchmark results in 45330.
- Remove keep_dims from ref impl signature.
- Use evaluate helpers to retrieve and normalize axes from host tensor.

Ticket: 37468, 45330

### CODE REVIEW

File: ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp [hunk 1]
```diff
@@ -21,44 +22,43 @@ namespace ngraph
...
-                      const Shape& in_shape,
-                      const AxisSet& reduction_axes,
-                      bool keep_dims)
+            void mean(const T* arg, T* out, const Shape& in_shape, const AxisSet& reduction_axes)
             {
-                auto out_shape = reduce(in_shape, reduction_axes, keep_dims);
-                CoordinateTransform output_transform(out_shape);
-                std::vector<T> cs(shape_size(out_shape));
+                constexpr bool dont_keep_dims_in_output = false;
+                auto out_shape = reduce(in_shape, reduction_axes, dont_keep_dims_in_output);
```
Comment (ilyachur) [inline ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp:28]: Do we need to have `keep_dims` in `reduce` reference implementation?

File: ngraph/core/src/op/reduce_mean.cpp [hunk 2]
```diff
@@ -68,6 +69,8 @@ bool op::v1::ReduceMean::evaluate(const HostTensorVector& outputs,
                                   const HostTensorVector& inputs) const
 {
     NGRAPH_OP_SCOPE(v1_ReduceMean_evaluate);
+    NGRAPH_CHECK(validate_host_tensor_vector(inputs, 2));
```
Comment (ilyachur) [inline ngraph/core/src/op/reduce_mean.cpp:72]: Looks like we don't use the second input. Am I right?

File: ngraph/core/src/op/reduce_mean.cpp [hunk 3]
```diff
@@ -68,6 +69,8 @@ bool op::v1::ReduceMean::evaluate(const HostTensorVector& outputs,
...
+    NGRAPH_CHECK(validate_host_tensor_vector(outputs, 1));
     return mean::evaluate_mean(inputs[0], outputs[0], get_reduction_axes(), get_keep_dims());
```
Comment (mitruska) [inline ngraph/core/src/op/reduce_mean.cpp:74]: Please update to reuse `get_normalized_axes_from_tensor` helper

Comment (ggalieroc) [reply ngraph/core/src/op/reduce_mean.cpp]: Done

File: ngraph/core/src/op/reduce_mean.cpp (hunk 2 above)
Comment (ggalieroc) [reply ngraph/core/src/op/reduce_mean.cpp:73]: After including helper function to retrieve axes from host tensor, this check makes sense

File: ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp (hunk 1 above)
Comment (ggalieroc) [reply ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp]: Please notice that `dont_keep_dims_in_output ` is a constant expression evaluated as false, that is used to indicate that we do not keep dims when using `reduce` method (As dims with value 1 do not influence in index calculation)

**Review by tomdol — APPROVED** [review]

File: ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp [hunk 4]
```diff
@@ -21,44 +22,43 @@ namespace ngraph
...
+                        index_to_count_map[out_idx]++;
                     }
 
                     if (is_finite(x) && is_finite(z))
                     {
-                        T& c = cs[output_transform.index(output_coord)];
+                        T& c = cs[out_idx];
                         T t = z + (x - c);
                         c = (t - z) - (x - c);
                         z = t;
```
Comment (mitruska) [inline ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp:64]: I know that this PR is rather about removing `index()` call so I approve that. 
But for me this piece of code is not easily understandable, 
to compare that `else` block is a simple sum.

**Review by mitruska — APPROVED** [review]

File: ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp (hunk 4 above)
Comment (ilyachur) [reply ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp:64]: I think while we are revising reference implementations we need to solve all issues.
At least I really don't understand the variables names (what does it mean? What is `z`, `x` and etc.?)

Comment (mitruska) [reply ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp:64]: Looks like: https://en.wikipedia.org/wiki/Kahan_summation_algorithm

Comment (ggalieroc) [reply ngraph/core/reference/include/ngraph/runtime/reference/mean.hpp]: I have added helper function with description. I hope this helps to understand.

File: ngraph/core/reference/include/ngraph/runtime/reference/sum.hpp [hunk 5]
```diff
@@ -18,28 +18,55 @@ namespace ngraph
...
-            template <typename T>
-            typename std::enable_if<std::is_floating_point<T>::value, bool>::type is_finite(T x)
+            namespace details
             {
-                return std::isfinite(x);
-            }
+                // Windows doesn't seem to like it if we directly use std::isfinite on integer
+                // types, so we will roll our own thing here.
+                template <typename T>
+                typename std::enable_if<std::is_floating_point<T>::value, bool>::type is_finite(T x)
```
[the remark this answers is not in the document]
Comment (ggalieroc) [reply ngraph/core/reference/include/ngraph/runtime/reference/sum.hpp:26]: In this case I don't think so. My understanding is that this provides separate function overloads and specialization for different type traits. This helper function is to return a bool to later decide how to perform summation of two elements, and the implementation is type specific

[the remark this answers is not in the document]
Comment (ggalieroc) [reply ngraph/core/reference/include/ngraph/runtime/reference/sum.hpp]: Done in the following commit: [3315f76](https://github.com/openvinotoolkit/openvino/pull/5806/commits/3315f769ff00e55bc0631e3ff7e78a5848b7d306)

**Review by ilyachur — APPROVED** [review]

**Pull request merged**